// discord.js-music-v11/index.js
const YoutubeDL = require('youtube-dl');
const ytdl = require('ytdl-core');
const fs = require("fs");

var setgame = fs.readFileSync("./game.json");

var jsonSetGame = JSON.parse(setgame);

/**
 * Takes a discord.js client and turns it into a music bot.
 * Thanks to 'derekmartinez18' for helping.
 * 
 * @param {Client} client - The discord.js client.
 * @param {object} options - (Optional) Options to configure the music bot. Acceptable options are:
 * 							prefix: The prefix to use for the commands (default '!').
 * 							global: Whether to use a global queue instead of a server-specific queue (default false).
 * 							maxQueueSize: The maximum queue size (default 20).
 * 							anyoneCanSkip: Allow anybody to skip the song.
 * 							clearInvoker: Clear the command message.
 * 							volume: The default volume of the player.
 */
module.exports = function (client, options) {
	// Get all options.
	let PREFIX = (options && options.prefix) || '!';
	let GLOBAL = (options && options.global) || false;
	let MAX_QUEUE_SIZE = (options && options.maxQueueSize) || 20;
	let DEFAULT_VOLUME = (options && options.volume) || 10;
	let ALLOW_ALL_SKIP = (options && options.anyoneCanSkip) || false;
	let CLEAR_INVOKER = (options && options.clearInvoker) || false;

	// Create an object of queues.
	let queues = {};

	// Catch message events.
	client.on('message', msg => {
		const message = msg.content.trim();
		if(msg.channel.type == 'dm') return false
		// Check if the message is a command.
		if (message.toLowerCase().startsWith(PREFIX.toLowerCase())) {
			// Get the command and suffix.
			const command = message.substring(PREFIX.length).split(/[ \n]/)[0].toLowerCase().trim();
			const suffix = message.substring(PREFIX.length + command.length).trim();

			// Process the commands.
			switch (command) {
				case 'play':
					return play(msg, suffix);
				case 'pl':
					return play(msg, suffix);
				case 'start':
					return play(msg, suffix);
				case 'skip':
					return skip(msg, suffix);
				case 's':
					return skip(msg, suffix);
				case 'queue':
					return queue(msg, suffix);
				case 'q':
					return queue(msg, suffix);
				case 'pause':
					return pause(msg, suffix);
				case 'p':
					return pause(msg, suffix);
				case 'resume':
					return resume(msg, suffix);
				case 'r':
					return resume(msg, suffix);
				case 'volume':
					return volume(msg, suffix);
				case 'vol':
					return volume(msg, suffix);
				case 'v':
					return volume(msg, suffix);
				case 'mute':
					return mute(msg, suffix);
				case 'm':
					return mute(msg, suffix);
				case 'unmute':
					return unmute(msg, suffix);
				case 'um':
					return unmute(msg, suffix);
				case 'leave':
					return leave(msg, suffix);
				case 'l':
					return leave(msg, suffix);
				case 'quit':
					return leave(msg, suffix);
				case 'stop':
					return leave(msg, suffix);
				case 'disconnect':
					return leave(msg, suffix);
				case 'clearqueue':
					return clearqueue(msg, suffix);
				case 'clearq':
					return clearqueue(msg, suffix);
				case 'c':
					return clearqueue(msg, suffix);
				case 'np':
					return nowplaying(msg);
				case 'nowplaying':
					return nowplaying(msg);
			}
			if (CLEAR_INVOKER) {
				msg.delete();
			}
		}
	});

	/**
	 * Checks if a user is an admin.
	 * 
	 * @param {GuildMember} member - The guild member
	 * @returns {boolean} - 
	 */
	function isAdmin(member) {
		return member.hasPermission("ADMINISTRATOR");
	}

	/**
	 * Checks if the user can skip the song.
	 * 
	 * @param {GuildMember} member - The guild member
	 * @param {array} queue - The current queue
	 * @returns {boolean} - If the user can skip
	 */
	function canSkip(member, queue) {
		if (ALLOW_ALL_SKIP) return true;
		else if (queue[0].requester === member.id) return true;
		else if (isAdmin(member)) return true;
		else return false;
	}

	/**
	 * Gets the song queue of the server.
	 * 
	 * @param {integer} server - The server id. 
	 * @returns {object} - The song queue.
	 */
	function getQueue(server) {
		// Check if global queues are enabled.
		if (GLOBAL) server = '_'; // Change to global queue.

		// Return the queue.
		if (!queues[server]) queues[server] = [];
		return queues[server];
	}

	/**
	 * The command for adding a song to the queue.
	 * 
	 * @param {Message} msg - Original message.
	 * @param {string} suffix - Command suffix.
	 * @returns {<promise>} - The response edit.
	 */
	function play(msg, suffix) {
		// Make sure the user is in a voice channel.
		if (msg.member.voiceChannel === undefined) return msg.channel.sendMessage('You\'re not currently in a voice channel.').then(m=>m.delete(15000));

		// Make sure the suffix exists.
		if (!suffix) return msg.channel.sendMessage('No video specified!').then(m=>m.delete(15000));

		// Get the queue.
		const queue = getQueue(msg.guild.id);

		// Check if the queue has reached its maximum size.
		if (queue.length >= MAX_QUEUE_SIZE) {
			return msg.channel.sendMessage('Maximum queue size reached!').then(m=>m.delete(15000));
		}

		// Get the video information.
		msg.channel.sendMessage('Searching...').then(response => {
			var searchstring = suffix
			if (!suffix.toLowerCase().startsWith('http')) {
				searchstring = 'gvsearch1:' + suffix;
			}

			YoutubeDL.getInfo(searchstring, ['-q', '--no-warnings', '--force-ipv4'], (err, info) => {
				// Verify the info.
				if (err || info.format_id === undefined || info.format_id.startsWith('0')) {
					return response.edit('Invalid video!').then(m=>m.delete(15000));
				}

				info.requester = msg.author.id;

				// Queue the video.
				response.edit('Enqueued: ' + "**" + info.title + "**" + " to be played. Position in queue: `" + (queue.length+1) + "` - added by "+ "**" + msg.author.username + "**").then(() => {
					queue.push(info);
					// Play if only one element in the queue.
					if (queue.length === 1) executeQueue(msg, queue);
				}).catch(console.log);
				response.delete(15000)
			});
		}).catch(console.log);
	}


	function nowplaying(msg) {
		const queue = getQueue(msg.guild.id);
		const video = queue[0];
		if (queue < 1) {
			msg.channel.sendMessage('I am currently not playing any music.').then(m=>m.delete(15000));
			return 
		}
		msg.channel.sendMessage('Now Playing: ' + "**" + video.title + "**").then(m=>m.delete(15000));
		client.user.setPresence({ game: { name: (video.title + " @michaelvuolo_"), url: "https://twitch.tv/presidentbingo", type: 1 } });
	}

	/**
	 * The command for skipping a song.
	 * 
	 * @param {Message} msg - Original message.
	 * @param {string} suffix - Command suffix.
	 * @returns {<promise>} - The response message.
	 */
	function skip(msg, suffix) {
		// Get the voice connection.
		const voiceConnection = client.voiceConnections.find(val => val.channel.guild.id == msg.guild.id);
		if (voiceConnection === null) return msg.channel.sendMessage('No music currently being played.').then(m=>m.delete(15000));

		// Get the queue.
		const queue = getQueue(msg.guild.id);
		const video = queue[0];		

		if (!canSkip(msg.member, queue)) return msg.channel.sendMessage('You cannot skip this as you didn\'t queue it.').then((response) => {
			response.delete(15000);
		});

		// Get the number to skip.
		let toSkip = 1; // Default 1.
		if (!isNaN(suffix) && parseInt(suffix) > 0) {
			toSkip = parseInt(suffix);
		}
		toSkip = Math.min(toSkip, queue.length);

		// Skip.
		queue.splice(0, toSkip - 1);

		// Resume and stop playing.
		const dispatcher = voiceConnection.player.dispatcher;
		if (voiceConnection.paused) dispatcher.resume();
		dispatcher.end();

		msg.channel.sendMessage('Skipped **' + video.title + '**!').then(m=>m.delete(15000));
	}

	/**
	 * The command for listing the queue.
	 * 
	 * @param {Message} msg - Original message.
	 * @param {string} suffix - Command suffix.
	 */
	function queue(msg, suffix) {
		// Get the queue.
		const queue = getQueue(msg.guild.id);

		// Get the queue text.
		const text = queue.map((video, index) => (
			"`" + (index+1) + "`" +'. ' + "**" + video.title + "**"
		)).join('\n');

		// Get the status of the queue.
		let queueStatus = 'Stopped';
		const voiceConnection = client.voiceConnections.find(val => val.channel.guild.id == msg.guild.id);
		if (voiceConnection !== null) {
			const dispatcher = voiceConnection.player.dispatcher;
			queueStatus = dispatcher.paused ? 'Paused' : 'Playing';
		}

		// Send the queue and status.
		msg.channel.sendMessage("Music is currently **" + queueStatus + "**.\n\n" + text).then(m=>m.delete(25000));
		if (queue.length < 1) {
			msg.channel.sendMessage("Use **+play <song name>** to play a song. :smiley:").then(m=>m.delete(25000));
		}
	}

	/**
	 * The command for pausing the current song.
	 * 
	 * @param {Message} msg - Original message.
	 * @param {string} suffix - Command suffix.
	 * @returns {<promise>} - The response message.
	 */
	function pause(msg, suffix) {
		// Get the voice connection.
		const voiceConnection = client.voiceConnections.find(val => val.channel.guild.id == msg.guild.id);
		if (voiceConnection === null) return msg.channel.sendMessage('No music is currently being played.').then(m=>m.delete(15000));

		if (!isAdmin(msg.member))
			return msg.channel.sendMessage('You are not authorized to use this.').then(m=>m.delete(15000));

		// Pause.
		msg.channel.sendMessage('â¸').then(m=>m.delete(15000));
		const dispatcher = voiceConnection.player.dispatcher;
		if (!dispatcher.paused) dispatcher.pause();
	}

	/**
	 * The command for leaving the channel and clearing the queue.
	 * 
	 * @param {Message} msg - Original message.
	 * @param {string} suffix - Command suffix.
	 * @returns {<promise>} - The response message.
	 */
	function leave(msg, suffix) {
		if (isAdmin(msg.member)) {
			const voiceConnection = client.voiceConnections.find(val => val.channel.guild.id == msg.guild.id);
			if (voiceConnection === null) return msg.channel.sendMessage('I am currently not in any channel.').then(m=>m.delete(15000));
			// Clear the queue.
			const queue = getQueue(msg.guild.id);
			queue.splice(0, queue.length);

			// End the stream and disconnect.
			voiceConnection.player.dispatcher.end();
			voiceConnection.disconnect();
			msg.channel.sendMessage("ðŸ‘‹");
			client.user.setPresence({ game: { name: jsonSetGame.game, type: 0 } });
		} else {
			msg.channel.sendMessage('You don\'t have permission to use that command!').then(m=>m.delete(15000));
		}
	}

	/**
	 * The command for clearing the song queue.
	 * 
	 * @param {Message} msg - Original message.
	 * @param {string} suffix - Command suffix.
	 */
	function clearqueue(msg, suffix) {
		if (isAdmin(msg.member)) {
			const queue = getQueue(msg.guild.id);

			queue.splice(0, queue.length);
			msg.channel.sendMessage('ðŸš®').then(m=>m.delete(15000));
		} else {
			msg.channel.sendMessage('You don\'t have permission to use that command!').then(m=>m.delete(15000));
		}
	}

	/**
	 * The command for resuming the current song.
	 * 
	 * @param {Message} msg - Original message.
	 * @param {string} suffix - Command suffix.
	 * @returns {<promise>} - The response message.
	 */
	function resume(msg, suffix) {
		// Get the voice connection.
		const voiceConnection = client.voiceConnections.find(val => val.channel.guild.id == msg.guild.id);
		if (voiceConnection === null) return msg.channel.sendMessage('No music is currently being played.').then(m=>m.delete(15000));

		if (!isAdmin(msg.member))
			return msg.channel.sendMessage('You are not authorized to use this.').then(m=>m.delete(15000));

		// Resume.
		msg.channel.sendMessage('â–¶').then(m=>m.delete(15000));
		const dispatcher = voiceConnection.player.dispatcher;
		if (dispatcher.paused) dispatcher.resume();
	}

	/**
	 * The command for changing the song volume.
	 * 
	 * @param {Message} msg - Original message.
	 * @param {string} suffix - Command suffix.
	 * @returns {<promise>} - The response message.
	 */
	function volume(msg, suffix) {
		// Get the voice connection.
		const voiceConnection = client.voiceConnections.find(val => val.channel.guild.id == msg.guild.id);
		if (voiceConnection === null) return msg.channel.sendMessage('No music is currently being played.').then(m=>m.delete(15000));

		if (!isAdmin(msg.member))
			return msg.channel.sendMessage('You are not authorized to use this.').then(m=>m.delete(15000));

		// Get the dispatcher
		const dispatcher = voiceConnection.player.dispatcher;
		if (suffix > 200 || suffix < 1 ) return msg.channel.sendMessage('Volume out of range!').then((response) => {
			response.delete(15000);
		});

		msg.channel.sendMessage("Volume set to " + suffix).then(m=>m.delete(15000));
		dispatcher.setVolume((suffix/100));
	}

	function mute(msg, suffix) {
		// Get the voice connection.
		const voiceConnection = client.voiceConnections.find(val => val.channel.guild.id == msg.guild.id);
		if (voiceConnection === null) return msg.channel.sendMessage('No music is currently being played.').then(m=>m.delete(15000));

		if (!isAdmin(msg.member))
			return msg.channel.sendMessage('You are not authorized to use this.').then(m=>m.delete(15000));

		// Get the dispatcher
		const dispatcher = voiceConnection.player.dispatcher;
		msg.channel.sendMessage("Volume muted.").then(m=>m.delete(15000));
		dispatcher.setVolume((0/100));
	}

	function unmute(msg, suffix) {
		// Get the voice connection.
		const voiceConnection = client.voiceConnections.find(val => val.channel.guild.id == msg.guild.id);
		if (voiceConnection === null) return msg.channel.sendMessage('No music is currently being played.').then(m=>m.delete(15000));

		if (!isAdmin(msg.member))
			return msg.channel.sendMessage('You are not authorized to use this.').then(m=>m.delete(15000));

		// Get the dispatcher
		const dispatcher = voiceConnection.player.dispatcher;
		msg.channel.sendMessage("Volume unmuted.").then(m=>m.delete(15000));
		dispatcher.setVolume((10/100));
	}

	/**
	 * Executes the next song in the queue.
	 * 
	 * @param {Message} msg - Original message.
	 * @param {object} queue - The song queue for this server.
	 * @returns {<promise>} - The voice channel.
	 */
	function executeQueue(msg, queue) {
		// If the queue is empty, finish.
		if (queue.length === 0) {
			msg.channel.sendMessage('ðŸ‘‹').then(m=>m.delete(15000));

			// Leave the voice channel.
			const voiceConnection = client.voiceConnections.find(val => val.channel.guild.id == msg.guild.id);
			if (voiceConnection !== null) return voiceConnection.disconnect();
		}

		new Promise((resolve, reject) => {
			// Join the voice channel if not already in one.
			const voiceConnection = client.voiceConnections.find(val => val.channel.guild.id == msg.guild.id);
			if (voiceConnection === null) {
				// Check if the user is in a voice channel.
				if (msg.member.voiceChannel) {
					msg.member.voiceChannel.join().then(connection => {
						resolve(connection);
					}).catch((error) => {
						console.log(error);
					});
				} else {
					// Otherwise, clear the queue and do nothing.
					queue.splice(0, queue.length);
					reject();
				}
			} else {
				resolve(voiceConnection);
			}
		}).then(connection => {
			// Get the first item in the queue.
			const video = queue[0];

			console.log('â— ' + msg.author.username + " is now playing " + video.webpage_url + " " + "(" + video.title + ")" + " in channel: " + msg.member.voiceChannel.name);

			// Play the video.
			msg.channel.sendMessage('Now playing in ' + msg.member.voiceChannel.name + ': ' + "**" + video.title + "**").then((response) => {
				let dispatcher = connection.playStream(ytdl(video.webpage_url, {filter: 'audioonly'}), {seek: 0, volume: (DEFAULT_VOLUME/100)});

				client.user.setPresence({ game: { name: (video.title + " @michaelvuolo_"), url: "https://twitch.tv/presidentbingo", type: 1 } });

				connection.on('error', (error) => {
					// Skip to the next song.
					console.log(error);
					queue.shift();
					executeQueue(msg, queue);
				});

				dispatcher.on('error', (error) => {
					// Skip to the next song.
					console.log(error);
					queue.shift();
					executeQueue(msg, queue);
				});

				dispatcher.on('end', () => {
					// Wait a second.
					setTimeout(() => {
						if (queue.length > 0) {
							// Remove the song from the queue.
							response.delete();
							queue.shift();
							client.user.setPresence({ game: { name: jsonSetGame.game, type: 0 } });
							// Play the next song in the queue.
							executeQueue(msg, queue);
						}
					}, 1000);
				});
			}).catch((error) => {
				console.log(error);
			});
		}).catch((error) => {
			console.log(error);
		});
	}
}

/**
 * Wrap text in a code block and escape grave characters.
 * 
 * @param {string} text - The input text.
 * @returns {string} - The wrapped text.
 */
function wrap(text) {
	return '```\n' + text.replace(/`/g, '`' + String.fromCharCode(8203)) + '\n```';
}
